homeassistant/components/nissan_leaf.py

```diff
@@ -0,0 +1,489 @@
+    async_dispatcher_connect, async_dispatcher_send)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_point_in_utc_time
+from homeassistant.util.dt import utcnow
+
+# If testing then use the following kinds of URLs
+#
+# REQUIREMENTS = ['https://github.com/filcole/pycarwings2/archive/master.zip'
+#                 '#pycarwings2']
+# REQUIREMENTS = ['https://test-files.pythonhosted.org/packages/7c/ad/
+#      ee27988357f1710ca9ced1a60263486415f054003ca6fa396922ca6b6bbf/
+#      pycarwings2-2.2.tar.gz'
+#                 '#pycarwings2']
+# REQUIREMENTS = ['file:///home/phil/repos/pycarwings2ve/pycarwings2/
+#      dist/pycarwings2-2.2.tar.gz'
+#                 '#pycarwings2']
+REQUIREMENTS = ['pycarwings2==2.3']
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'nissan_leaf'
+DATA_LEAF = 'nissan_leaf_data'
+
+DATA_BATTERY = 'battery'
+DATA_LOCATION = 'location'
+DATA_CHARGING = 'charging'
+DATA_PLUGGED_IN = 'plugged_in'
+DATA_CLIMATE = 'climate'
+DATA_RANGE_AC = 'range_ac_on'
+DATA_RANGE_AC_OFF = 'range_ac_off'
+
+CONF_NCONNECT = 'nissan_connect'
+CONF_INTERVAL = 'update_interval'
+CONF_CHARGING_INTERVAL = 'update_interval_charging'
+CONF_CLIMATE_INTERVAL = 'update_interval_climate'
+CONF_REGION = 'region'
+CONF_VALID_REGIONS = ['NNA', 'NE', 'NCI', 'NMA', 'NML']
+CONF_FORCE_MILES = 'force_miles'
+
+INITIAL_UPDATE = timedelta(seconds=15)
+MIN_UPDATE_INTERVAL = timedelta(minutes=2)
+DEFAULT_INTERVAL = timedelta(hours=1)
+DEFAULT_CHARGING_INTERVAL = timedelta(minutes=15)
+DEFAULT_CLIMATE_INTERVAL = timedelta(minutes=5)
+RESTRICTED_BATTERY = 2
+RESTRICTED_INTERVAL = timedelta(hours=12)
+
+MAX_RESPONSE_ATTEMPTS = 22
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(cv.ensure_list, [vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_REGION): vol.In(CONF_VALID_REGIONS),
+        vol.Optional(CONF_NCONNECT, default=True): cv.boolean,
+        vol.Optional(CONF_INTERVAL, default=DEFAULT_INTERVAL): (
+            vol.All(cv.time_period, vol.Clamp(min=MIN_UPDATE_INTERVAL))),
+        vol.Optional(CONF_CHARGING_INTERVAL,
+                     default=DEFAULT_CHARGING_INTERVAL): (
+                         vol.All(cv.time_period,
+                                 vol.Clamp(
+                                     min=MIN_UPDATE_INTERVAL))),
+        vol.Optional(CONF_CLIMATE_INTERVAL,
+                     default=DEFAULT_CLIMATE_INTERVAL): (
+                         vol.All(cv.time_period,
+                                 vol.Clamp(
+                                     min=MIN_UPDATE_INTERVAL))),
+        vol.Optional(CONF_FORCE_MILES, default=False): cv.boolean
+    })])
+}, extra=vol.ALLOW_EXTRA)
+
+LEAF_COMPONENTS = [
+    'sensor', 'switch', 'binary_sensor', 'device_tracker'
+]
+
+SIGNAL_UPDATE_LEAF = 'nissan_leaf_update'
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
```

Comment:
You need to review this function, it is way more complex then it needs to be.

look at the BMW_connect_drive component for an example

Reply:
@dgomes I'm unsure how to make this simpler.  Apart from this change request I think all other issues are resolved.